Python 3.10 support

Hello!

It would be great to get python 3.10 supported by this excellent package.

Thanks,
Rustam

Locally, there's one unit test failure I need to investigate. TBD re: system tests

```
(dev-3.10) ➜  python-bigquery git:(issue1006-python310) pytest tests/unit
====================================================== test session starts ======================================================
platform darwin -- Python 3.10.0, pytest-6.2.5, py-1.10.0, pluggy-1.0.0
rootdir: /Users/swast/src/github.com/googleapis/python-bigquery
collected 1889 items / 1 skipped / 1888 selected                                                                                

tests/unit/test__helpers.py sssssss...................................................................................... [  4%]
...................................................................                                                       [  8%]
tests/unit/test__http.py ......                                                                                           [  8%]
tests/unit/test__pandas_helpers.py ........................................................................sss........... [ 13%]
...............s...sssss........                                                                                          [ 15%]
tests/unit/test_client.py .........................................................................sss.s................. [ 20%]
......................................................................s.................................................. [ 26%]
....................                                                                                                      [ 27%]
tests/unit/test_create_dataset.py .............                                                                           [ 28%]
tests/unit/test_dataset.py ................................................................................               [ 32%]
tests/unit/test_dbapi__helpers.py ................................................................................        [ 36%]
tests/unit/test_dbapi_connection.py s..s.ss.s....                                                                         [ 37%]
tests/unit/test_dbapi_cursor.py ........................ssss...............................                               [ 40%]
tests/unit/test_dbapi_types.py ...........                                                                                [ 41%]
tests/unit/test_delete_dataset.py ...........                                                                             [ 41%]
tests/unit/test_encryption_configuration.py ..............                                                                [ 42%]
tests/unit/test_external_config.py ...........................................                                            [ 44%]
tests/unit/test_format_options.py ......                                                                                  [ 44%]
tests/unit/test_job_retry.py .........                                                                                    [ 45%]
tests/unit/test_list_datasets.py ....                                                                                     [ 45%]
tests/unit/test_list_jobs.py ........                                                                                     [ 46%]
tests/unit/test_list_models.py ..........                                                                                 [ 46%]
tests/unit/test_list_projects.py ....                                                                                     [ 46%]
tests/unit/test_list_routines.py ..........                                                                               [ 47%]
tests/unit/test_list_tables.py ..............                                                                             [ 48%]
tests/unit/test_opentelemetry_tracing.py sssssss                                                                          [ 48%]
tests/unit/test_query.py ................................................................................................ [ 53%]
...........................                                                                                               [ 55%]
tests/unit/test_retry.py ...................                                                                              [ 56%]
tests/unit/test_schema.py ............................................................................                    [ 60%]
tests/unit/test_signature_compatibility.py ..                                                                             [ 60%]
tests/unit/test_table.py ................................................................................................ [ 65%]
.....................s.......s.......s.Fssssss......s...s..ss......s..ssssssssssss..sss.ss............................... [ 71%]
............ssss                                                                                                          [ 72%]
tests/unit/enums/test_standard_sql_data_types.py ..s                                                                      [ 72%]
tests/unit/helpers/test_from_json.py ................................................                                     [ 75%]
tests/unit/job/test_base.py ...............................................................................               [ 79%]
tests/unit/job/test_copy.py ....................                                                                          [ 80%]
tests/unit/job/test_extract.py ................                                                                           [ 81%]
tests/unit/job/test_load.py .............................                                                                 [ 82%]
tests/unit/job/test_load_config.py ...................................................................................... [ 87%]
.......                                                                                                                   [ 87%]
tests/unit/job/test_query.py ...........................................................                                  [ 90%]
tests/unit/job/test_query_config.py ..........................                                                            [ 92%]
tests/unit/job/test_query_pandas.py ...........sssssss..sss..ss..sssssss                                                  [ 94%]
tests/unit/job/test_query_stats.py .........................                                                              [ 95%]
tests/unit/line_arg_parser/test_lexer.py .                                                                                [ 95%]
tests/unit/line_arg_parser/test_parser.py ...........                                                                     [ 96%]
tests/unit/line_arg_parser/test_visitors.py .                                                                             [ 96%]
tests/unit/model/test_model.py ....................                                                                       [ 97%]
tests/unit/model/test_model_reference.py ...........                                                                      [ 97%]
tests/unit/routine/test_routine.py .........................                                                              [ 99%]
tests/unit/routine/test_routine_argument.py ......                                                                        [ 99%]
tests/unit/routine/test_routine_reference.py ...........                                                                  [100%]

=========================================================== FAILURES ============================================================
______________________________ TestRowIterator.test_to_arrow_ensure_bqstorage_client_wo_bqstorage _______________________________

self = <tests.unit.test_table.TestRowIterator testMethod=test_to_arrow_ensure_bqstorage_client_wo_bqstorage>

    @unittest.skipIf(pyarrow is None, "Requires `pyarrow`")
    def test_to_arrow_ensure_bqstorage_client_wo_bqstorage(self):
        from google.cloud.bigquery.schema import SchemaField
    
        schema = [
            SchemaField("name", "STRING", mode="REQUIRED"),
            SchemaField("age", "INTEGER", mode="REQUIRED"),
        ]
        rows = [
            {"f": [{"v": "Alice"}, {"v": "98"}]},
            {"f": [{"v": "Bob"}, {"v": "99"}]},
        ]
        path = "/foo"
        api_request = mock.Mock(return_value={"rows": rows})
    
        mock_client = _mock_client()
        mock_client._ensure_bqstorage_client.return_value = None
        row_iterator = self._make_one(mock_client, api_request, path, schema)
    
        tbl = row_iterator.to_arrow(create_bqstorage_client=True)
    
        # The client attempted to create a BQ Storage client, and even though
        # that was not possible, results were still returned without errors.
>       mock_client._ensure_bqstorage_client.assert_called_once()

tests/unit/test_table.py:2570: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

_mock_self = <MagicMock name='mock._ensure_bqstorage_client' spec='function' id='4987555360'>

    def assert_called_once(_mock_self):
        """assert that the mock was called only once.
        """
        self = _mock_self
        if not self.call_count == 1:
            msg = ("Expected '%s' to have been called once. Called %s times.%s"
                   % (self._mock_name or 'mock',
                      self.call_count,
                      self._calls_repr()))
>           raise AssertionError(msg)
E           AssertionError: Expected '_ensure_bqstorage_client' to have been called once. Called 0 times.

/usr/local/Caskroom/miniconda/base/envs/dev-3.10/lib/python3.10/site-packages/mock/mock.py:891: AssertionError
==================================================== short test summary info ====================================================
FAILED tests/unit/test_table.py::TestRowIterator::test_to_arrow_ensure_bqstorage_client_wo_bqstorage - AssertionError: Expecte...
========================================== 1 failed, 1796 passed, 93 skipped in 30.84s ==========================================
```

--- a/.kokoro/presubmit/snippets-2.7.cfg
+++ b/.kokoro/presubmit/snippets-2.7.cfg
@@ -1,7 +1,0 @@
-# Format: //devtools/kokoro/config/proto/build.proto
-
-# Only run this nox session.
-env_vars: {
-    key: "NOX_SESSION"
-    value: "snippets-2.7"
-}

--- a/.kokoro/presubmit/snippets-3.10.cfg
+++ b/.kokoro/presubmit/snippets-3.10.cfg
@@ -3,5 +3,5 @@
 # Only run this nox session.
 env_vars: {
     key: "NOX_SESSION"
-    value: "prerelease_deps-3.8"
+    value: "snippets-3.10"
 }

--- a/.kokoro/presubmit/system-3.10.cfg
+++ b/.kokoro/presubmit/system-3.10.cfg
@@ -3,5 +3,5 @@
 # Only run this nox session.
 env_vars: {
     key: "NOX_SESSION"
-    value: "prerelease_deps-3.8"
+    value: "system-3.10"
 }

--- a/noxfile.py
+++ b/noxfile.py
@@ -69,7 +69,12 @@ def default(session, install_extras=True):
         constraints_path,
     )
 
-    install_target = ".[all]" if install_extras else "."
+    if install_extras and session.python == "3.10":
+        install_target = ".[bqstorage,pandas,tqdm,opentelemetry]"
+    elif install_extras:
+        install_target = ".[all]"
+    else:
+        install_target = "."
     session.install("-e", install_target, "-c", constraints_path)
 
     session.install("ipython", "-c", constraints_path)
@@ -153,7 +158,11 @@ def system(session):
     # Data Catalog needed for the column ACL test with a real Policy Tag.
     session.install("google-cloud-datacatalog", "-c", constraints_path)
 
-    session.install("-e", ".[all]", "-c", constraints_path)
+    if session.python == "3.10":
+        extras = "[bqstorage,pandas,tqdm,opentelemetry]"
+    else:
+        extras = "[all]"
+    session.install("-e", f".{extras}", "-c", constraints_path)
     session.install("ipython", "-c", constraints_path)
 
     # Run py.test against the system tests.

--- a/noxfile.py
+++ b/noxfile.py
@@ -27,7 +27,7 @@ BLACK_VERSION = "black==19.10b0"
 BLACK_PATHS = ("docs", "google", "samples", "tests", "noxfile.py", "setup.py")
 
 DEFAULT_PYTHON_VERSION = "3.8"
-SYSTEM_TEST_PYTHON_VERSIONS = ["3.8"]
+SYSTEM_TEST_PYTHON_VERSIONS = ["3.8", "3.10"]
 UNIT_TEST_PYTHON_VERSIONS = ["3.6", "3.7", "3.8", "3.9", "3.10"]
 CURRENT_DIRECTORY = pathlib.Path(__file__).parent.absolute()
 

--- a/noxfile.py
+++ b/noxfile.py
@@ -28,7 +28,7 @@ BLACK_PATHS = ("docs", "google", "samples", "tests", "noxfile.py", "setup.py")
 
 DEFAULT_PYTHON_VERSION = "3.8"
 SYSTEM_TEST_PYTHON_VERSIONS = ["3.8"]
-UNIT_TEST_PYTHON_VERSIONS = ["3.6", "3.7", "3.8", "3.9"]
+UNIT_TEST_PYTHON_VERSIONS = ["3.6", "3.7", "3.8", "3.9", "3.10"]
 CURRENT_DIRECTORY = pathlib.Path(__file__).parent.absolute()
 
 # 'docfx' is excluded since it only needs to run in 'docs-presubmit'

--- a/samples/geography/noxfile_config.py
+++ b/samples/geography/noxfile_config.py
@@ -22,7 +22,12 @@
 
 TEST_CONFIG_OVERRIDE = {
     # You can opt out from the test for specific Python versions.
-    "ignored_versions": ["2.7"],
+    "ignored_versions": [
+        "2.7",
+        # TODO: Enable 3.10 once there is a geopandas/fiona release.
+        # https://github.com/Toblerity/Fiona/issues/1043
+        "3.10",
+    ],
     # An envvar key for determining the project id to use. Change it
     # to 'BUILD_SPECIFIC_GCLOUD_PROJECT' if you want to opt in using a
     # build specific Cloud project. You can also use your own string

--- a/samples/geography/requirements.txt
+++ b/samples/geography/requirements.txt
@@ -24,14 +24,12 @@ importlib-metadata==4.8.1
 libcst==0.3.21
 munch==2.5.0
 mypy-extensions==0.4.3
-numpy==1.19.5; python_version < "3.7"
-numpy==1.21.2; python_version > "3.6"
 packaging==21.0
 pandas==1.1.5; python_version < '3.7'
-pandas==1.3.2; python_version >= '3.7'
+pandas==1.3.4; python_version >= '3.7'
 proto-plus==1.19.2
 protobuf==3.18.0
-pyarrow==5.0.0
+pyarrow==6.0.0
 pyasn1==0.4.8
 pyasn1-modules==0.2.8
 pycparser==2.20
@@ -43,7 +41,7 @@ pytz==2021.1
 PyYAML==5.4.1
 requests==2.26.0
 rsa==4.7.2
-Shapely==1.7.1
+Shapely==1.8.0
 six==1.16.0
 typing-extensions==3.10.0.2
 typing-inspect==0.7.1

--- a/samples/magics/requirements.txt
+++ b/samples/magics/requirements.txt
@@ -1,4 +1,3 @@
-google-cloud-bigquery==2.27.1
 google-cloud-bigquery-storage==2.9.0
 google-auth-oauthlib==0.4.6
 grpcio==1.41.0

--- a/samples/magics/requirements.txt
+++ b/samples/magics/requirements.txt
@@ -3,10 +3,10 @@ google-cloud-bigquery-storage==2.9.0
 google-auth-oauthlib==0.4.6
 grpcio==1.41.0
 ipython==7.16.1; python_version < '3.7'
-ipython==7.17.0; python_version >= '3.7'
+ipython==7.29.0; python_version >= '3.7'
 matplotlib==3.3.4; python_version < '3.7'
-matplotlib==3.4.1; python_version >= '3.7'
+matplotlib==3.5.0rc1; python_version >= '3.7'
 pandas==1.1.5; python_version < '3.7'
-pandas==1.3.2; python_version >= '3.7'
-pyarrow==5.0.0
+pandas==1.3.4; python_version >= '3.7'
+pyarrow==6.0.0
 pytz==2021.1

--- a/samples/snippets/requirements.txt
+++ b/samples/snippets/requirements.txt
@@ -1,4 +1,3 @@
-google-cloud-bigquery==2.27.1
 google-cloud-bigquery-storage==2.9.0
 google-auth-oauthlib==0.4.6
 grpcio==1.41.0

--- a/samples/snippets/requirements.txt
+++ b/samples/snippets/requirements.txt
@@ -3,10 +3,10 @@ google-cloud-bigquery-storage==2.9.0
 google-auth-oauthlib==0.4.6
 grpcio==1.41.0
 ipython==7.16.1; python_version < '3.7'
-ipython==7.17.0; python_version >= '3.7'
+ipython==7.29.0; python_version >= '3.7'
 matplotlib==3.3.4; python_version < '3.7'
 matplotlib==3.4.1; python_version >= '3.7'
 pandas==1.1.5; python_version < '3.7'
-pandas==1.3.2; python_version >= '3.7'
-pyarrow==5.0.0
+pandas==1.3.4; python_version >= '3.7'
+pyarrow==6.0.0
 pytz==2021.1

--- a/setup.py
+++ b/setup.py
@@ -127,6 +127,7 @@ setuptools.setup(
         "Programming Language :: Python :: 3.7",
         "Programming Language :: Python :: 3.8",
         "Programming Language :: Python :: 3.9",
+        "Programming Language :: Python :: 3.10",
         "Operating System :: OS Independent",
         "Topic :: Internet",
     ],
@@ -135,7 +136,7 @@ setuptools.setup(
     namespace_packages=namespaces,
     install_requires=dependencies,
     extras_require=extras,
-    python_requires=">=3.6, <3.10",
+    python_requires=">=3.6, <3.11",
     include_package_data=True,
     zip_safe=False,
 )

--- a/setup.py
+++ b/setup.py
@@ -59,7 +59,7 @@ extras = {
     ]
     + pyarrow_dep,
     "geopandas": ["geopandas>=0.9.0, <1.0dev", "Shapely>=1.6.0, <2.0dev"],
-    "pandas": ["pandas>=0.23.0"] + pyarrow_dep,
+    "pandas": ["pandas>=0.24.2"] + pyarrow_dep,
     "bignumeric_type": pyarrow_dep,
     "tqdm": ["tqdm >= 4.7.4, <5.0.0dev"],
     "opentelemetry": [

